src/meta/src/stream/stream_graph/fragment.rs

```diff
@@ -331,6 +336,262 @@ impl StreamFragmentEdge {
+        panic!("Sink has more than 1 input: {:?}", sink_node.input);
+    };
+    let PbNodeBody::StreamScan(scan) = stream_scan_node.node_body.as_ref().unwrap() else {
+        return Err(anyhow!(
+            "expect PbNodeBody::StreamScan but got: {:?}",
+            stream_scan_node.node_body
+        )
+        .into());
+    };
+    let stream_scan_type = PbStreamScanType::try_from(scan.stream_scan_type).unwrap();
+    if stream_scan_type != PbStreamScanType::ArrangementBackfill {
+        return Err(anyhow!(
+            "unsupported stream_scan_type for auto refresh schema: {:?}",
+            stream_scan_type
+        )
+        .into());
+    }
+    let [merge_node, _batch_plan_node] = stream_scan_node.input.as_slice() else {
+        panic!(
+            "the number of StreamScan inputs is not 2: {:?}",
+            stream_scan_node.input
+        );
+    };
+    let NodeBody::Merge(_) = merge_node.node_body.as_ref().unwrap() else {
+        return Err(anyhow!(
+            "expect PbNodeBody::Merge but got: {:?}",
+            merge_node.node_body
+        )
+        .into());
+    };
+    Ok(())
+}
+
+pub fn rewrite_refresh_schema_sink_fragment(
+    original_sink_fragment: &Fragment,
+    sink: &PbSink,
+    newly_added_columns: &[ColumnCatalog],
+    upstream_table: &PbTable,
+    upstream_table_fragment_id: FragmentId,
+    id_generator_manager: &IdGeneratorManager,
+) -> MetaResult<(Fragment, Vec<PbColumnCatalog>, Option<PbTable>)> {
+    let mut new_sink_columns = sink.columns.clone();
+    fn extend_sink_columns(
+        sink_columns: &mut Vec<PbColumnCatalog>,
+        new_columns: &[ColumnCatalog],
+        get_column_name: impl Fn(&String) -> String,
+    ) {
+        let next_column_id = sink_columns
+            .iter()
+            .map(|col| col.column_desc.as_ref().unwrap().column_id + 1)
+            .max()
+            .unwrap_or(1);
+        sink_columns.extend(new_columns.iter().enumerate().map(|(i, col)| {
+            let mut col = col.to_protobuf();
+            let column_desc = col.column_desc.as_mut().unwrap();
+            column_desc.column_id = next_column_id + (i as i32);
+            column_desc.name = get_column_name(&column_desc.name);
+            col
+        }));
+    }
+    extend_sink_columns(&mut new_sink_columns, newly_added_columns, |name| {
+        name.clone()
+    });
+
+    let mut new_sink_fragment = clone_fragment(original_sink_fragment, id_generator_manager);
+    let sink_node = &mut new_sink_fragment.nodes;
+    let PbNodeBody::Sink(sink_node_body) = sink_node.node_body.as_mut().unwrap() else {
+        return Err(anyhow!("expect PbNodeBody::Sink but got: {:?}", sink_node.node_body).into());
+    };
+    let [stream_scan_node] = sink_node.input.as_mut_slice() else {
+        panic!("Sink has more than 1 input: {:?}", sink_node.input);
+    };
+    let PbNodeBody::StreamScan(scan) = stream_scan_node.node_body.as_mut().unwrap() else {
+        return Err(anyhow!(
```

Comment:
It is possible that there is a StreamExchange between StreamSink and StreamScan when the upstream table's pk is different from the sink upsert pk.

Reply:
For iceberg sink with partition by, there would be an additional ProjectNode here, but I think we don't need to handle it in this PR. Just remind that in the future implementation.